http/src/v0/root_files.rs

```diff
@@ -0,0 +1,60 @@
+use crate::v0::support::unshared::Unshared;
+use crate::v0::support::{with_ipfs, StreamResponse, StringError};
+use ipfs::{Ipfs, IpfsTypes};
+use libipld::cid::Codec;
+use serde::Deserialize;
+use std::convert::TryFrom;
+use warp::{path, query, Filter, Rejection, Reply};
+
+#[derive(Debug, Deserialize)]
+pub struct CatArgs {
+    // this could be an ipfs path
+    arg: String,
+    offset: Option<u64>,
+    length: Option<u64>,
+    // timeout: Option<?> // added in latest iterations
+}
+
+pub fn cat<T: IpfsTypes>(
+    ipfs: &Ipfs<T>,
+) -> impl Filter<Extract = impl Reply, Error = Rejection> + Clone {
+    path!("cat")
+        .and(with_ipfs(ipfs))
+        .and(query::<CatArgs>())
+        .and_then(cat_inner)
+}
+
+async fn cat_inner<T: IpfsTypes>(ipfs: Ipfs<T>, args: CatArgs) -> Result<impl Reply, Rejection> {
+    use crate::v0::refs::{walk_path, IpfsPath};
+    use ipfs::unixfs::{ll::file::FileReadFailed, TraversalFailed};
+
+    let path = IpfsPath::try_from(args.arg.as_str()).map_err(StringError::from)?;
+
+    let range = match (args.offset, args.length) {
+        (Some(start), Some(len)) => Some(start..(start + len)),
+        (Some(_start), None) => todo!("need to abstract over the range"),
+        (None, Some(len)) => Some(0..len),
+        (None, None) => None,
+    };
+
+    // FIXME: this is here until we have IpfsPath back at ipfs
+
+    let (cid, _, _) = walk_path(&ipfs, path).await.map_err(StringError::from)?;
```

Comment:
nit: I think the pattern `let (cid, ..)` would work here; it's not much, but makes life easier when adding or removing a tuple field later on

Reply:
I'd prefer to get an error in case I find a need to change it yet again. Hopefully no need to change it :)